File: sem3/map/Lab/toylanguage_C#/ToyLanguage/ToyLanguage/Models/Expressions/VariableExpression.cs
using System;
using ToyLanguage.Models.Interfaces;
using ToyLanguage.Models.States;

namespace ToyLanguage.Models.Expressions
{
    public class VariableExpression : IExpression
    {
        private readonly string _variableName;

        public VariableExpression(string variableName)
        {
            _variableName = variableName;
        }
        
        public int Eval(ProgramState state)
        {
            if (!state.SymbolTable.ContainsKey(_variableName))
            {
                throw new Exception("Invalid variable name!");
            }

            return state.SymbolTable[_variableName];
        }

        public override string ToString()
        {
            return _variableName;
        }
    }
}

File: sem3/map/Exam/GUI_ToyLanguageInterpreter/src/Interpreter/Models/Expressions/VariableExpression.java
package Interpreter.Models.Expressions;

import Interpreter.Models.ExceptionFactory;
import Interpreter.Models.Expression;
import Interpreter.Models.ProgramState.ProgramState;
import Interpreter.Utils.MyIDictionary;

public class VariableExpression extends Expression {
    private String variableName;

    public VariableExpression(String variableName) {
        this.variableName = variableName;
    }

    @Override
    public int eval(ProgramState state) throws Exception {
        MyIDictionary<String, Integer> symTable = state.getSymTable();

        if (!symTable.contains(this.variableName)) {
            throw ExceptionFactory.create("variable", "variable not declared!");
        }

        return symTable.getValue(this.variableName);
    }

    @Override
    public String toString() {
        return variableName;
    }
}
